### ISSUE TITLE
Counter KDF HMAC-SHA2-256 Key derivation key length (keyIn) always 128 bits

### ISSUE THREAD

**Issue by Sv3n** [usnistgov/ACVP#1013]
The testvectors that are returned while having registered for "KDF", using `"kdfMode": "counter"` and `"macMode": "HMAC-SHA2-256"` always look somewhat like this:

```json
        {
            "tgId": 1,
            "keyOutLength": 64,
            "kdfMode": "counter",
            "macMode": "HMAC-SHA2-256",
            "counterLength": 32,
            "counterLocation": "before fixed data",
            "testType": "AFT",
            "tests": [
                {
                    "tcId": 1,
                    "keyIn": "469A022AAE25C84123E849D930652406"
                },
                {
                    "tcId": 2,
                    "keyIn": "80321980048F76D66ADF36DD10D3E469"
                }
            ]
        }
```

In both tests (and also all other test groups) the "keyIn" parameter is 128 bits. Two issues with that:

1. Given the underlying algorithm is "HMAC-SHA2-256", I would have expected a default of 256 bits for the key derivation key length. All the downloadable test vectors in countermode for HMAC-SHA-256 use a 256 bit value for KI here, for example: https://csrc.nist.gov/Projects/Cryptographic-Algorithm-Validation-Program/Key-Derivation
2. It might make sense to be able to register the supported lengths for "keyIn", if the intention is to test more than 1 length for each macMode (if that is done, then it might be worth considering avoiding ambiguity with `supportedLengths`, since it is not immediately clear this refers to the supported output key lengths).

Full registration for reference:

```
      {
        "algorithm": "KDF",
        "revision": "1.0",
        "prereqVals": [
          {
            "algorithm": "SHA",
            "valValue": "same"
          },
          {
            "algorithm": "HMAC-SHA2-256",
            "valValue": "same"
          }
        ],
        "capabilities": [
          {
            "kdfMode": "counter",
            "macMode": [
              "HMAC-SHA2-256"
            ],
            "supportedLengths": [
              {
                "min": 64,
                "max": 1024,
                "increment": 64
              },
              2048,
              3072,
              4096
            ],
            "fixedDataOrder": [
              "before fixed data"
            ],
            "counterLength": [
              32
            ],
            "supportsEmptyIv": false,
            "requiresEmptyIv": false
          }
        ]
      }
```

**Comment by Sv3n** [comment]
Are there plans to pick up this issue in the near future? Being able to test a 256 bit KDF at its intended security strength through ACVP is something that my company would like to do soon. If it is not available, then we need to begin on planning to add in work-arounds.

**Comment by Kritner** [comment]
Thanks for your patience @Sv3n, been mostly trying to focus on bug fixes and performance improvements prior to taking on new features.  I did generate some sample vectors using a 256 bit key in (with the intention of using it against sha2-256), you're welcome to run them for testing purposes.

[prompt.json.txt](https://github.com/usnistgov/ACVP/files/5966867/prompt.json.txt) (actually a json file but github won't allow a .json to be attached).

Regarding the request, this is something I often struggle with, and wanted to get some outside perspectives.

> Given the underlying algorithm is "HMAC-SHA2-256", I would have expected a default of 256 bits for the key derivation key length. All the downloadable test vectors in countermode for HMAC-SHA-256 use a 256 bit value for KI here, for example: https://csrc.nist.gov/Projects/Cryptographic-Algorithm-Validation-Program/Key-Derivation

There are a few possibilities to serve this request:

* Change the default `keyIn` length to match the digest size of the algorithm
  * this seems to make a lot of sense, and it looks like it should have been this way from the start especially given the previous quote in this comment
  * Would this be a breaking change for anyone? I wouldn't *think* so, since the HMAC implementation itself should be able to support a key length of "any" size and pad it or truncate it to the digest size of the algorithm either way.
    * though in this case, you have an IUT that *only* support a `keyIn` of the digest size of the underlying algorithm, which I guess could be for hardware/speed specific reasons.
* New testing revision as to not "break" the existing testing by changing the `keyIn` length on something that has already been used in production
  * This touches on something that could quickly get out of hand if new testing revisions were introduced each time we needed to accommodate an IUT that didn't support our testing... I mention this more as a "general case" scenario, rather than this specific case, which I'll attempt to elaborate on below.***

Personally, I'm hoping that just changing the `keyIn` lengths to match the digest size works for everyone, it means no new testing revision, and seems like it should be a transparent change to most folks.

-----

***Given that IUTs may have some restrictions that we don't account for, like supporting some random length that (to us) may not be supported by our testing (again not this digest size for the `keyIn` situation, but in a more general sense), 

We (or at least I) often struggle with the potential granularity of what is supported vs not when coming up with registration properties.  Obviously the more properties you have, the more control the implementers have over the vectors produced, but also the more potential for registration property explosion and confusion on how to go about using the system.

We want it to be *flexible* but we also want it to be *simple* - the more properties we add to the registration the more flexible it potentially becomes, but the less simple; if that makes sense.

If we *were* to support providing "valid key lengths" per chosen HMAC function (as an example) that would obviously give the most granularity to the users of the system, but also makes the registrations that more complex.

I'm curious what you (and others) feel about the above.  If you take the first iteration of KAS testing (1.0) in comparison to the current version (56Ar3) you'll see that the registrations are much simpler, at the cost of some granularity.  Was that an appropriate consolidation of properties? Looking even at the new KAS testing I see opportunities to perhaps further consolidate in the next testing revision; but I'm not sure if these consolidations are a welcome change for ease of use, or an unwelcome one due to the loss of the specificity on a per property basis.

**Comment by Sv3n** [comment]
Thank you for the elaborate response! I will try to address the points you mentioned:

> About the 256 bit keyIn json file

We will give this a try! We always test our implementation against internal references vectors as well, so I do not expect problems, but going though all the motions to check formatting and other consistency problems is valuable regardless.

> (...) Would this be a breaking change for anyone?

Looking at this purely from my own perspective: in the same way there exists an incompatibility with our IUT and the 128-bit keys, a reverse incompatibility can exist for anyone who built their implementation having only 128-bit keys in mind when the default switches to 256.

Your assumption that this may be hardware / speed related is correct. Some context on why parameter support is narrowed down in our case (which may translate to others as well):

A generic HMAC algorithm indeed support all allowed key lengths, given the padding / truncation specification. In practice, available resources limit the size, e.g.:

  1. SHA needs to keep track of the message length, and the counter has a limited size
  2. Keys and other data have to be transported into the DUT: buffering space may be limited anywhere along this path (either fundamentally, or by specific design choices (a streaming style communication interface deals better with this compared to a fully buffered one))
  3. The HMAC interface (the hardware equivalent of the API) may have limited expressiveness: if you can't tell it to read a key of 128 bits (or if you can't tell it to begin padding after 128 bits), then that size becomes unusable.

The reasons for imposing these limits along the way typically fall under one of these categories:

- Time (costs): All combinations of inputs an implementation supports have to be tested, consuming time (development, testing, documentation). Both in the CAVP context, but also more broadly to ensure the hardware is correct.
- Area / power: Practically everything has a hardware cost (in terms of area / gate count). Larger counters, or wider communication buses all translate into more gates.
- Security: Each additional allowed parameter adds another possible attack angle. Think buffer / counter overflows, or potential side channels that get opened, broader fuzzing ranges.

The hardware that sequences the steps to implement the KDF may be limited as well, along similar lines (counters, fixed_data generation / buffering / API). On top of that, we are dealing with a IUT that does not offer all of the algorithms that it uses as service to its users: instead we are dealing with a dedicated test interface that only exists to demonstrates that we are internally doing everything according to spec. In the field, during normal use, we only work at 256-bit security strength (and there is no way to do anything else): from that point of view, setting up the support for parameters that deviate from regular use is pure overhead.

Many of these points also translate to software implementations to a some extent (replace area by code size). I think software might be more flexible in some sense, where the overhead of supporting something extra, just for testing purposes might be smaller, but this can also be a pain depending on the software's abstraction level (higher difficulty options: optimized assembly, ROM code).

> We want it to be flexible but we also want it to be simple - the more properties we add to the registration the more flexible it potentially becomes, but the less simple; if that makes sense.

It's a tough one: the broader your registration support is, the more specific everyone's implementation can be (and only you get to deal with all the associated issues (testing / documentation / compute power) :)). **Sensible and explicit defaults** might be a partial solution: in this case it is not specified what `keyIn` length to expect, and supporting every possibility is not practical. So specifying a default size (e.g. the digest size) already helps. If that works for you as an ACVP user, then you are done (the simple case). If it does not, then you can go for a more specific registration, which then is more complex, but also required at that point.

The downside of taking this approach is that this may grandfather in some of the "not so intuitive" defaults in the spirit of not breaking backwards compatibility, to the point where a new revision is required eventually to straighten it out, and potentially having to support 2 versions in parallel as a result. I am aware that the ACVP specs are technically still in Draft status, and with that somewhat fluid, but at the same time they are already being used by labs and module makers. Ideally there comes a time at which everything is stable enough, such that a chip that is designed to pass tests right now, is still able to do that after it has been fabricated (at least a couple of months later).

Even through I am (selfishly) in favor of changing the (implicit) default `keyIn` length here, it can be quite costly for others, because a change like that is effectively "removing" support for something that previously worked: and that can have a pretty large fall-out on a hardware development process, for example. So changing the registration options is friendlier in that regard. My suggestion instead would therefore be to add an optional keyInLength parameter. When it is not specified, use the current default of 128. This makes the tool backward compatible, and also still allows for future expansion.

**Comment by Kritner** [comment]
Apologies on the delay again, and thank you for the response that was quite insightful.  I feel like the comments in this issue can likely apply to a few of the other issues that have come in recently.  Overridable defaults on properties that we don't currently even "take in" as properties, is likely something we'll have to cover at some point in the future.

That being said, I plan on changing the default values `keyIn` lengths to match the output length of their respective aux function, 256 for sha256, 512 for sha512, etc.  This seems to have been the original intent when having ported it over from CAVS, so it would hopefully have minimal impact.

I'll try to get this in for the next release.

**Comment by EdSmith-Viasat** [comment]
@Kritner : Your proposed solution of changing the default values keyIn lengths to match the output length of their respective "aux function" (the standard uses the term "PRF") doesn't match the standard and wouldn't be the correct solution for all implementations.  

The SP 800-108 standard makes no mention of the size of KeyIn, but does define the PRF as an approved HMAC or CMAC.  The output lengths of the approved HMAC and CMAC are 128, 224, 256, 384, and 512 bits.  

Typical symmetric keys sizes are 128, 192, or 256 bits.  

We have an implementation that uses a 256 bit KeyIn and and HMAC-SHA-384 as the aux function.  While your proposed solution won't require a significant modification to our implementation, it won't be testing how our implementation is utilized.  

I recommend KeyIn be specified as 128, 192, or 256 bits and not be dependent of the output size of the PRF.

**Comment by livebe01** [comment]
@EdSmith-Viasat, curious about your implementation. If you're using HMAC-SHA-384 with a 256-bit key and an L of 384, you now need to use a 384-bit key to get a 256-bit security strength. Can you explain/comment on why an implementor would do this? If you have time to share, great. If not, no problem. Just trying to learn.

**Comment by EdSmith-Viasat** [comment]
@livebe01, we use an L of 256.  According to table 3 in SP 800-56 part 1, rev 5, an HMAC-SHA-384 algorithm with a 256 bit Key Derivation Key (KDK, or keyIn) can produce a derived key at the same security strength as the input KDK.  

Please correct me if I'm wrong with my interpretation of SP 800-56.

**Comment by livebe01** [comment]
Sorry @EdSmith-Viasat, can you clarify the SP reference? Are you referring to 800-56C, Rev2, table 3?

**Comment by EdSmith-Viasat** [comment]
@livebe01, My mistake, I meant to refer to table 3 in SP 800-57 part 1, rev 5.

**Comment by livebe01** [comment]
Okay, thanks. Tracking now.

Table 3 in SP 800-57 part 1, rev 5 shows that SHA-384 for key derivation/HMAC/etc. will provide >= 256 bits of security strength. There's a note above the table that reads, "Note that in the case of HMAC and KMAC, which require keys, the estimated security strength assumes that the length and entropy used to generate the key are at least equal to the security strength." So I think I see what you're saying in that you can use a 256-bit key for HMAC-SHA-384.

My thinking on why there might be an issue with your implementation is as follows. You're implementing an SP 800-108 KDF in counter mode. From section 5.1 of SP 800-108:
1) n =  ⎡256/384⎤= 1
2) ...
3) result(0):= ∅.
4) For i = 1 to 1, do
a) K(1) := PRF (KI, [i]2 || Label || 0x00 || Context || [L]2)
b) result(1) := result(0) || K(1).
5) Return: KO := the leftmost 256 bits of result(1)

I assume that the result of PRF (KI, [i]2 || Label || 0x00 || Context || [L]2) would be a 384 bit value.

In SP 800-108, section 7.1 the authors make the assumption, when examining the security strength of the KDF, that "the key derivation key, KI, is the only unknown input to the KDF."

So the 256 bits of security strength (perhaps you could think of it as 256 bits of entropy) seem to me to get spread out over the 384 bit output of step 4a. And then in step 5, the 384 bit value (containing 256 bits of entropy) is truncated to 256 bits. So, in my reasoning, the 256 bit value that's produced now has a security strength (or entropy) of <256 bits due to that truncation. 

My reasoning could definitely be off, but that's what I was thinking :).

**Comment by EdSmith-Viasat** [comment]
@livebe01, I follow your reasoning.  

Based on SP 800-108, section 7.1, the security strength of the derived key is: [(256/384) * e] = 0.67e ; where 'e' is the entropy per bit of the DRBG used to produce the key derivation key.  
If the DRBG produces bits with 1 bit of entropy, then the resulting security strength of the key will be 170 bits.

**Comment by Kritner** [comment]
@EdSmith-Viasat introduced a `customKeyInLength` that be used in the `capability` object for KDF, this will allow for the overriding of our default keyInLength which currently matches the output length of the PRF (for hashes only - since AES/TDES keyIn lengths cannot be changed)

Related PR https://github.com/usnistgov/ACVP/pull/1182, this change is already on demo.

**Comment by Kritner** [comment]
This change has been released to prod in release [v1.1.0.18](https://github.com/usnistgov/ACVP-Server/releases/tag/v1.1.0.18).
